tests/test_meta_commands.py

```diff
@@ -746,12 +746,108 @@ def test_foreach_object_completion(hlwm):
 
 
 def test_foreach_flags_completion(hlwm):
-    flag_name = '--filter-name='
-    assert flag_name in hlwm.complete(['foreach'], partial=True)
-    assert flag_name in hlwm.complete(['foreach', 'X'], partial=True)
-    assert flag_name in hlwm.complete(['foreach', 'X', 'tags.'], partial=True)
-    # do not complete flags when the first command token appeared
-    assert flag_name not in hlwm.complete(['foreach', 'X', 'tags.', 'echo'], partial=True)
+    for flag_name in ['--filter-name=', '--unique ', '--recursive ']:
+        assert flag_name in hlwm.complete(['foreach'], partial=True)
+        assert flag_name in hlwm.complete(['foreach', 'X'], partial=True)
+        assert flag_name in hlwm.complete(['foreach', 'X', 'tags.'], partial=True)
+        # do not complete flags when the first command token appeared
+        assert flag_name not in hlwm.complete(['foreach', 'X', 'tags.', 'echo'], partial=True)
+
+
+@pytest.mark.parametrize('running_clients_num', [5])
+def test_foreach_unique(hlwm, running_clients, running_clients_num):
+    cmd_prefix = ['foreach', 'C', 'clients.']
+    cmd_suffix = ['sprintf', 'WINID_ATTR', '%c.winid', 'C',
+                  'substitute', 'WINID', 'WINID_ATTR', 'echo', 'WINID']
+    uniq = hlwm.call(cmd_prefix + ['--unique'] + cmd_suffix).stdout.splitlines()
+    duplicates = hlwm.call(cmd_prefix + cmd_suffix).stdout.splitlines()
+
+    focus_winid = hlwm.attr.clients.focus.winid()
+
+    # the only difference between the two is the focused client
+    # which is represented twice in 'clients.'
+    assert sorted(duplicates) == sorted(uniq + [focus_winid])
+
+
+def test_foreach_recursive_breadth_first(hlwm):
+    hlwm.call('split bottom')
+    hlwm.call('split left')
+
+    def object_paths():
+        return hlwm.call(['foreach', 'X', '', '--recursive', 'echo', 'X']) \
+                   .stdout.splitlines()
+
+    without_client = object_paths()
+
+    # and put the a client in the frame leaf:
+    winid, _ = hlwm.create_client()
+
+    with_client = object_paths()
+
+    # since recursive is in breadth-first order,
+    # the client's parent_frame should be visited before
+    # the deeply nested tags.focus.tiling.(root|focused_frame)...
+
+    found_in_without_client = False
+    for tag_id in ['0', 'focus']:
+        frame_object = f'tags.{tag_id}.tiling.focused_frame'
+        assert frame_object not in with_client
+        if frame_object in without_client:
+            found_in_without_client = True
+
+    assert found_in_without_client, \
+        "The frame must be listed under tags. before client creation"
```

Comment:
replace . with ,?

Reply:
actually, I meant the object `tags.`, I've simply replaced it with `'tags'`

Reply:
Aaah, thanks for clearing that up! :)